Add support for the Dynamic(xyz) wallet library

rdig opened the pull request:
This PR introduces CDapp support for multiple wallet types, trough the use of the [Dynamic](https://www.dynamic.xyz/) wallet library / SaaS.

As with `web3-onboard` which we were using previously, it allows wallet abstraction through use of proprietary interfaces and classes removing the need to interact with the specific wallets directly.

Due to it being inherently a SaaS, it requires a always-on connection and a API key, meaning it makes running it locally kind of problematic _(but not impossible)_

Besides the expect and available wallet types, it also has support for custodial wallets _(Embedded wallets in Dynamic parlance)_, which means that you can use a OAuth logic _(Google, Github)_ and have a wallet created automatically for you. This is useful for less web3 savy users.

![Screenshot from 2025-01-23 18-03-44](https://github.com/user-attachments/assets/1d94a895-f0cf-4fa6-9ce6-b91e593e82f9)

Improving from last wallet implementation iteration, I've took the opportunity to fix live network and wallet address changes which didn't use to work properly.

Now, we have cases to handle both changing of the user's wallet network, as well as changing the wallet's current address

https://github.com/user-attachments/assets/e1a4e584-0c65-49eb-8229-a28e29924403

## Milestones

This is part of two milestones related to Wallet support, this being Milestone 0, _(M0)_ from now on.

It is also to be considered the _feature branch_ as, even after deploying into production I have **no plans of merging this back into `master`** until M1 will be merged back into this.

M1 deals with refactoring `web3-onboard` to function as a fallback for the Dynamic wallet system in cases where this is either ran locally, the connection to the dynamic servers fails, or someone wants to run the CDapp trully decentralized.

![Screenshot from 2025-01-23 17-58-15](https://github.com/user-attachments/assets/3a7fe3ae-b3ac-4d6b-9aa0-91a29e8e7c8a)
![Screenshot from 2025-01-23 17-58-25](https://github.com/user-attachments/assets/366ca147-db24-4972-8ca9-2f072a776078)

It will also aid in local development as it will bring back software wallets which can auto sign transactions and messages easing the time it takes to test various paths _(this is the main reason for not wanting to merge this)_

## Rabbi Wallet Type

One the most requested new wallet type seems to be [Rabbi](https://rabby.io/) for some reason, however it does not come without it's issues.

First, it does not have support out of the box for testnets, meaning something like local development and Arbitrum Sepolia will have to be manually added by the user _(mainnets like Arbitrum are fine though)_

![Screenshot from 2025-01-23 18-09-20](https://github.com/user-attachments/assets/0dc6d7ed-fea9-4578-a6c7-476556d94840)

However, one of it's biggest issues is the fact that **if** the connect popover it spans _(which seems to be different from all others)_ looses focus, like via moving your mouse cursor, it will cancel the connection giving you a instant failure message. 

It's made even worse by users running Window Managers which tent to focus on hover, making this problem even worse:

https://github.com/user-attachments/assets/bbf120e0-9204-4a1f-9528-e4dc2d482429

It is not only something that we noticed, as it seems to be tracked via Rabbi's github issues: RabbyHub/Rabby#1920 _(which seems to be opened for more than a year at this point)_

However, any other popover that Rabbi opens _(signing a message, signing a transaction, etc)_ will work just fine out of focus:

https://github.com/user-attachments/assets/47177f4f-4cf3-4737-b258-7c23d15d2fb2

## Custodial Wallet (Embedded Wallet)

Custodial wallets are different not only because the user is not in control of them, but also since they don't have their own interface, so one must be provided for them _(to change networks, send funds, sign, etc)_

Due to this, we now have an extra entry in the User Hamburger menu which essentially gives the user access to the wallet interface whenever they require it

![Screenshot from 2025-01-23 18-02-10](https://github.com/user-attachments/assets/7a364c4d-0fc3-4b71-887f-e247a6513c3f)

## Brave Browser/Wallet/Bitski

Two things I must highlight regarding using the Brave Browser, and inherently it's wallet to connect and use the CDapp.

For one, the Brave Wallet is actually the Bitski under the hood _(which was recently acquired by Phantom)_, but with a different interface, meaning both "Brave Wallet" and "Bitski" will show up as installed when using Brave

![Screenshot from 2025-01-28 23-40-55](https://github.com/user-attachments/assets/4ede028d-a55d-4a76-831a-e5fbf1ea957e)

Second, developing locally via the Brave/Bitski Wallet directly is currently not possible as due to the security features of Brave, it expects a secure connection to both the frontend and the RPC endpoint, while our local dev environment runs on a standard one.

![Screenshot from 2025-01-28 23-39-37](https://github.com/user-attachments/assets/cc6b57e8-c5c2-4dec-aa4e-273eaa29def7)

Using another wallet, like Metamask, through Brave should be possible though.

## Implementation Quirks

There's two issues that stood out as more "problematic" during this implementation.

One being that, even though the documentation clearly lists it out https://docs.dynamic.xyz/react-sdk/hooks/usedynamicevents it will not trigger the listener(s) for wallet change events _(`primaryWalletChanged`)_. Due to this, I had to manually set up a makeshift listener by making use of `useEffect`.

The second issue is due to the high level that the `DynamicContextProvider` context needs to be in the rendering hierarchy, it does not have access to the CDapp's light/dark themes from context. For this, I had to tap back into `localStorage` to fetch the state, that is already set in there by `PageThemeContextProvider`. This is basically double dipping. 

A third is that due to the CDapp only supporting `ethers@5.x` and dynamic only supporting `ethers@6.x` I had to use an older package, which means there's a bunch of dependency overrides in the tree, which generates a lot of npm install warnings

![Screenshot from 2025-01-23 19-06-50](https://github.com/user-attachments/assets/b014839f-e438-47ea-8c14-3c576354080b)

## Live Debugging

For live debugging you can enable the `__DYNAMIC_WALLET_DEBUG_ENABLED__` variable in your `localStorage` which will
- enable the Dynamic native wallet control view
- more output in the console related to the internal wallet state

## Running Locally

As I mentioned above, this is really not intended to be run locally for development, even though it can be.

To achieve this, you'll need to add a new variable to your `.env.local.secrets`: _(ping me for the sandbox key)_
```bash
DYNAMIC_ENV_ID=
```
After which you'll need to run `npm run prepare`.

Once this is set up, you can run you local environment as before, however you'll only have access to non-dev wallets, so I suggest adding the ganache development wallets in Metamask and use those. 

They can be found at: http://localhost:3006/ganache-accounts.json

Alternatively, generate new wallets from any of the wallet types available, and use one of the development wallets to send funds into it, however, this will only be effecting for the current testing session.

## Testing

Due to it working the same both locally, on QA, or in Production, the testing procedure is the same, just with the network being changed _(so I suggest, for ease of use, to actually test it in a live environment -- QA)_

Prerequisites if running this locally:
- Update your `.env.local.secrets` file with the new `DYNAMIC_ENV_ID` env variable
- Run `npm run prepare`
- Run the dev environment `npm run dev`
- Create test data `node scripts/create-data.js`
- Run the frontent `npm run frontend`

Actual testing:
1. Connect with one of your preferred wallets _(the modal will list all that are supported, and if not installed locally, will point you to a URL that'll help you install)_
2. If the wallet has a "dapp connection" mechanism _(Metamask, Rabbi)_ make sure you get a popup that ask you to connect
3. After, make sure you get prompted to sign the authentication message
4. Send a transaction _(without metratransactions enabled)_
5. Sign a message _(either send a metatransaction, or vote/reveal)_
6. With the wallet connected, manually change the network -- ensure the network changes in the CDapp
7. With the wallet connected, manually change the address -- ensure the address change is reflected in the CDapp _(make sure the address you're changing to is connected to the CDapp though -- you are using a wallet that implements the "app conections" pattern)_

## Thanks

As with all my recent PRs, special thanks go out to @jakubcolony for helping the issue with click event bubbling when using Embedded wallets, and @area for helping debug Rabbi's _"weirdness"_

Resolves #992 
Resolves #1022 
Resolves #1037 
Resolves #2796
Resolves #4184
Resolves #4177
Resolves #4188
Resolves #4176
Resolves #4206
Resolves #4211
Resolves #4213
Resolves JoinColony/colonyCDappAuthProxy#35
Resolves #4219
Resolves #4227

arrenv reviewed (dismissed):
@rdig exciting to see this one up, thank you!

I know this is still in draft, but I couldn't help myself.


1. Is there a reason you went with Sequence over other options, for example, Thirdweb?

2. I tested the flow of Sequence and it seems to differ in the app compared to their demo.
In the app I had to follow these steps:
    1. Click to open the connect modal.
    2. Click on Sequence (which is not clear what options it provides i.e. email signup).
    3. Then a popup appears where I can select the sign in option.
    4. Then I can enter email.
    5. Then I get screen with the unique code.
    6. In my email I then get a link, which opens a new window.
    7. I then get a page where I need to go back to the pop up window, then grab the unique code, then go to the new window from the email and enter a code.
  ![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/697e8fa0-ae1c-4a0c-a409-2e6eb05575d1)
    8. I then get a confirmation screen.
  ![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/92918b14-fd08-40f5-953a-9991a57a6ac4)
    9. I then need to return to the app, if I wait about 30 sec, the modal and popup auto close. Otherwise I need to manually close it.
    10. I then need to click connect again.
    11. Select Sequence.
    12. Confirm the terms.
    13. Click connect.
    14. Connected.

     However, testing their wallet connector here - https://0xsequence.github.io/kit/?mode=waas - I had the following steps:
    1. Click to open the connect modal.
    2. Enter email.
    3. Check email which includes code.
    4. Return to app and enter code.
    5. Connected.

3. When using Sequence, I am connected to Polygon (assuming this is due to it not supporting local dev), and I am not able to switch networks, reloading the page switched the network - 
![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/e0aaf8e8-c19b-41bd-ad49-6a5a5af7646f)

4. For all except Dev wallets, I was not able to disconnect the wallet, until I reloaded the page.

5. For all except Dev wallets, I was not able to connect again, until I reloaded the page.

6. I was not able to get Brave wallet to connect with the following error.

![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/3afd7a9f-884b-4ca7-92d3-b2d15db5823a)

7. Coinbase wallet connected, but was connected to Ethereum and I was not able to change the network. Reloading the page switched the network in the UI.

8. Trying to test creating a Colony with a Coinbase wallet, resulted in it still thinking it was on Ethereum and I received the error to suggest the same.
![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/3759eeac-33a2-4df7-b972-6c13d84e0c2b)

![image](https://github.com/JoinColony/colonyCDapp/assets/33682027/675a6c5c-aa90-441d-85fe-9b0e7afb8d3d)

arrenv commented:
Thank you for continuing to work on this @rdig, it will be great to get this up and running. Are the two that we have now the only ones that we can add?

Is there something I need to set up first?

Whenever I try to connect with any other wallet, tried Brave, Walletconnect (Paper), Walletconnect (Zerion), it triggers Metamask with a different address, if I sign, it connects with the address I want but that does not exist in Metamask, but I am not able to do anything after that.


![brave-connecting](https://github.com/user-attachments/assets/b60e7a3b-4ca1-42ab-8bf6-dea4c1f935b9)

I have tried a number of things but have not been able to test more beyond that.

rdig commented:
So that's one of the interactions I wanted us to chat about, and it's the nature of browser injected wallets.

Since they're basically the same under the hood, or more specifically using the same injected provider, they'll kinda clobber each other, if you're going to try with all of them enabled at once.

And even more specifically, like in your case, if one is already "connected" to the dapp, like your Metamask. And when I say connect I'm not talking about the mechanism to connect your wallet to the dapp through our interface, but to connect your wallet to the dapp internally in metamask _(or whatever other wallet)_. These are two distinctive things

This can even work the other way around. Disconect Metamask (from Metamask's interface), connect Zerion, and then each time you'll try to login with Metamask, Zerion will pop up instead.

Actually, out of all Zerion handles this better since it has a "Metamask mode", meaning it will disable itself when you're using Metamask as to not cause conflicts.

In closing, yeah, this won't be pretty to solve

arrenv commented:
@rdig Thank you for the details. Seems like a complicated issue.

I have been testing Safe's implementation of Onboard.js and I am able to connect to all wallets as you would expect, despite having Metamask installed and connected. I can switch between wallets without any problems. Is there anything they are doing special or differently to be able to achieve this result?

rdig commented:
As discussed, the issue seems to be our auto-login mechanism, so that's going to be looked into in order to fix the reported issues

rdig committed:
Update: `@web3-onboard` to latest release

rdig committed:
Fix: `AppContextProvider` types after @web3-onboard updates

rdig committed:
Add: introduce `dynamic` wallet framework

rdig committed:
Chore: update package log after rebase

rdig committed:
Fix: `estimateGasCost` tx options update logic

Only wait for the tx options update mutation to succeed
*if one was sent* otherwise just proceed with the tx

rdig committed:
Fix: `estimateGasCost` tx options update logic

Only wait for the tx options update mutation to succeed
*if one was sent* otherwise just proceed with the tx

rdig committed:
Chore: update and fix package/package lock

rdig committed:
Add: dynamic env file var

rdig committed:
Add: `UserNavigation` show debug dynamic wallet controls

rdig committed:
Add: `WalletConnectedTopMenu` manage embedded wallet button

rdig committed:
Fix: `getGasPrices` dynamic transport RPC call

rdig committed:
Remove: un-needed `isFullWallet` checks

rdig committed:
WIP: refactor `AppContext` to support dynamic wallets

rdig committed:
Refactor: RetryProvider to extend dynamic transport

rdig committed:
Add: `Entry` initialize dynamic wallet context

rdig committed:
Add: `Entry` introduce top level app dark mode

rdig committed:
Chore: update phospor icons to match types

rdig committed:
Fix: various type complaints brought on by dynamic

rdig committed:
Fix: Dockerfile add required runtimes for the new deps

rdig committed:
Add: enable dynamic logs on non-local environments

rdig committed:
Fix: 'estimateGasCost` after rebase

jakubcolony committed:
Fix: Do not close action sidebar when clicking within dynamic wallet modal

jakubcolony committed:
Fix: Also do not close the sidebar when clicking on other Dynamic modals

jakubcolony committed:
Try: Disable Dynamic library shadow DOM

jakubcolony committed:
Revert "Fix: Also do not close the sidebar when clicking on other Dynamic modals"

This reverts commit 4e118b56ec5955ba066ca4d68510106c94e2bd00.

jakubcolony committed:
Revert "Fix: Do not close action sidebar when clicking within dynamic wallet modal"

This reverts commit adcc201c5206836107e584e1e175d9e7fddd0f27.

jakubcolony committed:
Fix: somewhat propert solution for Dynamic modals closing action sidebar

rdig committed:
Fix: show embedded wallet controls as long as one connected

rdig committed:
Fix: `AppContextProvider` handle wallet network change

rdig committed:
Add: `AppContextProvider` handle wallet out of sync

rdig committed:
Add: `AppContextProvider` handle wallet address change

rdig committed:
Add: `AppContextProvider` handle wallet connect/sign cancel

rdig committed:
Refactor: `Entry` use hook for dark move state

rdig committed:
Add: minor dynamic wallet styles overrides

rdig committed:
Remove: onboard wallet saga leftover comments

rdig committed:
Change: enable Dynamic wallet usage tracking

rdig committed:
Add: dynamic wallet modal light mode tag bg override

arrenv reviewed (changes requested):
@rdig Very cool to see this one up. These are a few things I experienced.

1. I like the addition of the Manage wallet option for embedded wallets. Can we change the text of if to "Manage wallet".

    ![image](https://github.com/user-attachments/assets/6227886b-8d28-4f13-bd94-1a4ecc6eb001)

2. While in the manage embedded wallet modal, I used the inbuilt "Logout" feature, which logged me out of Dynamic but did not actually log me out of the app. I can still navigate around, go to my user account area and see the wallet is connected in the header.

    ![image](https://github.com/user-attachments/assets/516e58b8-a0ee-46e8-8791-21c5816fa069)
    
    ![image](https://github.com/user-attachments/assets/41e1ed9b-079c-443e-9a58-87e93ee3f581)


3. It would be great to update the OTP email that gets sent out, I reached out to Dynamic support and they said it is possible, however have not provided details as to how. Can you see anything in the Dynamic dashboard for it?

    ![image](https://github.com/user-attachments/assets/cb3d5a34-3c64-48e6-a28d-4305d5408fd9)


4.  If you reject the login in your wallet, you get a little stuck, this seems like more of an issue depending on the connection option you choose, for example, Brave requires you to reload the app in order to resolve it.

    ![image](https://github.com/user-attachments/assets/8e63f0d6-9474-4940-91f3-bdc5213d3deb)

    ![image](https://github.com/user-attachments/assets/fd202ffd-a7a8-45c2-a6fa-0c4fde76cd1d)

5. There seems to be a few cases where connecting a wallet for the first time where authentication did not work. It does not happen in every case but it has happened a couple of times. To resolve it I had to disconnect, reload the app and connect again with the same wallet.

    ![image](https://github.com/user-attachments/assets/d9991c69-87ac-4a56-8fb1-10d6e628e1a0)
    
    ![image](https://github.com/user-attachments/assets/50bf3c26-9897-4a1b-a9fe-9547e2c95fbd)

6. I ran into this issue when trying to login via Google, however, after disconnecting and trying again it seemed to work the second time.

    ![image](https://github.com/user-attachments/assets/533cb7d0-4a03-40b9-9fec-ca58ad5fef3b)


7. Not really sure what you can do about this, but just pointing it out. I have Bitski wallet showing as installed, although I definitely do not have it installed anywhere. My assumption is that it is related to Brave in some way, if anyone else can confirm they see the same.

    ![image](https://github.com/user-attachments/assets/d0811288-8a39-4041-b36e-d5b6a7f1dfe3)


8. Are there any options in the Dynamic dashboard to change the domain or text here so that it shows "Colony" instead?

    ![image](https://github.com/user-attachments/assets/e222ce50-9a4c-4a9a-a799-6e6eb8b941ec)

9. When using oAuth, e.g. Google, there seems to be no redirect that happens if you don't click continue within a minute, despite a successful login.

    ![image](https://github.com/user-attachments/assets/e20076bc-ee6b-4175-bb42-8ef7129e6690)

    ![image](https://github.com/user-attachments/assets/70067721-8210-430f-9fb7-95430db134be)

rdig commented:
@arrenv 

> 1. I like the addition of the Manage wallet option for embedded wallets. Can we change the text of if to "Manage wallet".

I'll change it, sure, but this implies that other wallets can be "managed"

> 2. While in the manage embedded wallet modal, I used the inbuilt "Logout" feature, which logged me out of Dynamic but did not actually log me out of the app. I can still navigate around, go to my user account area and see the wallet is connected in the header.

Hopefully they have a event for that, and it actually works, otherwise I can't do anything

> 3. It would be great to update the OTP email that gets sent out, I reached out to Dynamic support and they said it is possible, however have not provided details as to how. Can you see anything in the Dynamic dashboard for it?

I did not see anything in the dashboard related to this, mostly about adding names and logos _(perhaps adding a logo will render it in the email)_

> 4. If you reject the login in your wallet, you get a little stuck, this seems like more of an issue depending on the connection option you choose, for example, Brave requires you to reload the app in order to resolve it.

I'll check brave, but your example there with Metamask is not stuck, you just have to press back.

This will be different for most every wallet, and sooner or later one of them is going to do something different that we can't account for

> 5. There seems to be a few cases where connecting a wallet for the first time where authentication did not work. It does not happen in every case but it has happened a couple of times. To resolve it I had to disconnect, reload the app and connect again with the same wallet.

I need replication steps for this, the devtools screenshot just shows a graphql query that's unauthenticated, which is expected if the wallet failed to connect

> 6. I ran into this issue when trying to login via Google, however, after disconnecting and trying again it seemed to work the second time.

We can't really control that


> 7. Not really sure what you can do about this, but just pointing it out. I have Bitski wallet showing as installed, although I definitely do not have it installed anywhere. My assumption is that it is related to Brave in some way, if anyone else can confirm they see the same.

We can't really control that

> 8. Are there any options in the Dynamic dashboard to change the domain or text here so that it shows "Colony" instead?

This is not a dynamic thing, it's an OAuth thing, and the redirect URL has to be on dynamic's servers, so that URL will always show like that

> 9. When using oAuth, e.g. Google, there seems to be no redirect that happens if you don't click continue within a minute, despite a successful login.

I don't understand this one, how is there a successful login, if you don't click continue ?

rdig commented:
> 5. There seems to be a few cases where connecting a wallet for the first time where authentication did not work. It does not happen in every case but it has happened a couple of times. To resolve it I had to disconnect, reload the app and connect again with the same wallet.

I know what's going on with this. One of my recent changes, after it's deployed will fix it

rdig commented:
> 7. Not really sure what you can do about this, but just pointing it out. I have Bitski wallet showing as installed, although I definitely do not have it installed anywhere. My assumption is that it is related to Brave in some way, if anyone else can confirm they see the same.

Still nothing we can do about it, but turns out, you do have in fact Bitski installed, since that's basically Brave's wallet, they just re-skkined it

But you can use it to connect, and the Brave interface will show up :)

arrenv commented:
@rdig 

1. Thank you, the menu item did not seem to be showing for me if it was not an embedded wallet, so that should be fine.

2. Nice, looks like we have an event https://docs.dynamic.xyz/react-sdk/hooks/usedynamicevents#usedynamicevents

9. After 1 minute, I did click Continue and it authenticated me, it just didn't know where to redirect me and resulted in that white screen.

rdig commented:
@arrenv 

> 2. Nice, looks like we have an event https://docs.dynamic.xyz/react-sdk/hooks/usedynamicevents#usedynamicevents

That's why I said, "if", since their docs lie. Check the PR description as I have a similar issue with `primaryWalletChanged`

> 3. After 1 minute, I did click Continue and it authenticated me, it just didn't know where to redirect me and resulted in that white screen.

I'll try to replicate this, but again, not our systems, so we have little recourse

rdig committed:
Refactor: `WalletConnectedTopMenu` manage wallet copy

rdig committed:
Fix: `AppContextProvider` handle embedded wallet logout

rdig committed:
Add: automatically switch wallet network upon connect

rdig committed:
Fix: `AppContext` wallet recover after traumatic event

Fixes #4184

rdig committed:
Fix: `LandingPage` wallet trigger multiple loading states

Fixes #4177

jakubcolony committed:
Refactor action sidebar before close callbacks, fix closing on Dynamic modal click

rdig commented:
_NOTE: The description was updated with a section on Brave_

rdig committed:
Fix: remove dynamic wallet shadow dom disable

Fixes #4176

rdig committed:
Fix: prevent redirect to landing if wallet will autoconnect

Fixes: #4206

rdig committed:
Fix: `AppContextProvider` handle dynamic local storage

rdig committed:
Refactor: dynamic wallet css overrides for non shadow dom

rdig committed:
Refactor: dynamic CSS overrides as strings

rdig committed:
Remove: `UserMenu` "Get Started" menu entry

Fixes #4213

rdig committed:
Add: `StepCreateUser` auto insert email from oauth2

rdig committed:
Fix: `AppContextProvider` fix autologin logic for embedded

Fixes #4206

rdig committed:
Experimental: app is in loading state until proxy auth sig

rdig committed:
Refactor: `AppContext` only switch chain if different

There's a Metamask issue where it wouldn't connect if you asked it
to switch to the same chain

rdig committed:
Fix: `AppContext` reauthenticate wallet on address change

rdig committed:
Add: `AppContext` better wallet auto chain change debug log
